### PR TITLE
23 tests for hot swappable numpy backends demo use

### LINKED ISSUE #23: Tests for hot-swappable numpy backends, demo use
Adding a jax backend option in `katsu.katsu_math.py` via the `BackendShim` class, and revamping `katsu.mueller` functions to handle Jax’s sharp bits would be a great addition for differentiable Mueller calculus. 

An example on how to add compatibility to functions in `katsu.mueller` with a simplified version of the `linear_polarizer` function

```python
def linear_polarizer(a, shape=None):
    “""docstring
    """

    # returns array of zeros
    M = _empty_mueller(shape)

    ones = np.ones_like(a)
    cos2a = np.cos(2 * a)
    sin2a = np.sin(2 * a)

    # fist row
    if np.__name__ == ‘jax.numpy’:
        M.at[..., 0, 0].set(ones)
        
    else:
        M[..., 0, 0] = ones

    return M
```

[Heres the link to all of Jax’s sharp bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#array-updates-x-at-idx-set-y)

### PR DESCRIPTION
**Pull request by kenjim21**
added jax backend compatibility for mueler functions and corresponding testing

### CODE REVIEW

**Comment by Jashcraf** [comment]
Ah looks like adding Jax as an optional dependency and not explicitly testing it hurts code coverage. I think that's fine for now, but just writing a note here to consider a way of structuring a test file to maintain the code coverage. Maybe removing it from the automated codecov reporting or just configuring the tests on the remote machine

File: katsu/mueller.py [hunk 1]
```diff
@@ -363,47 +448,91 @@ def decompose_diattenuator(M, normalize=False):
...
-        diattenuation_vector = M[..., 0, 1:] / T
+    if np.__name__ == "jax.numpy":
+        if M.ndim > 2:
+            diattenuation_vector = M.at[..., 0, 1:].divide(T.at[..., np.newaxis])
+        else:
+            diattenuation_vector = M.at[..., 0, 1:].divide(T)
 
-    D = np.sqrt(np.sum(diattenuation_vector * diattenuation_vector, axis=-1))
-    mD = np.sqrt(1 - D**2)
+        D = np.sqrt(np.sum(np.matmul(diattenuation_vector, diattenuation_vector), axis=-1))
```
Comment (Jashcraf) [inline katsu/mueller.py]: It looks like an element wise multiplication was replaced with a matmul. Does this end up doing a dot product because it does so in the last axis?

Comment (Jashcraf) [reply katsu/mueller.py]: Yeah I think this might need to be reverted to an element-wise multiplication, but let me check the source material first.

Comment (kenjim21) [reply katsu/mueller.py]: I just tried reverting the change and uncommenting the test for it. still getting a failure unfortunately (though this could still be because of the eigen values thing)

**Comment by kenjim21** [comment]
For the codecov we can def look at adding all the possible autodiff libraries that are being supported to a special designation for testing so that while it's not on the requirements.txt, it'll be install for the automated testing

File: katsu/mueller.py (hunk 1 above)
Comment (Jashcraf) [reply katsu/mueller.py]: So decompose_diattenuator doesn’t rely on eigenvalue computation, which suggests that maybe its something significant

Comment (Jashcraf) [reply katsu/mueller.py]: i’ll try poke around on my machine

Comment (Jashcraf) [reply katsu/mueller.py]: image
works with the version you have with matmul

Comment (Jashcraf) [reply katsu/mueller.py]: image
Works for me with elementwise multiplication in `decompose_diattenuator`

Comment (kenjim21) [reply katsu/mueller.py]: is that with the test uncommented?

Comment (Jashcraf) [reply katsu/mueller.py]: If I did it right I think i just commited the change

Comment (kenjim21) [reply katsu/mueller.py]: okay, let me push uncommenting the test and see if it passes on the workflow

Comment (Jashcraf) [reply katsu/mueller.py]: Yeah for decompose_diattenuator, not for decompose_depolarizer

Comment (kenjim21) [reply katsu/mueller.py]: wait we still have it not required so it won't check it regardless. can you test on your computer if the tests work now?

Comment (Jashcraf) [reply katsu/mueller.py]: decompose_depolarizer fails if I uncomment it

**Comment by Jashcraf** [comment]
Oh man did it run with the test uncommented?

**Comment by kenjim21** [comment]
Screenshot 2024-08-16 180213

This is the margin of difference between the matrices when I print them

**Comment by Jashcraf** [comment]
So that value in the lower-left is concerning, but not huge - can you make an issue to investigate this difference? For now I think this is fine for the PR

**Comment by Jashcraf** [comment]
3…2…1…

LGTM!

**Comment by kenjim21** [comment]
Ya i'll add an issue. it looks like for any of the non zero values minus the top row there's its almost consistently off at around the 6th decimal or further
